sympy/algebras/quaternion.py

```diff
@@ -244,14 +254,13 @@ def add(self, other):
 
         # If q2 is a number or a sympy expression instead of a quaternion
         if not isinstance(q2, Quaternion):
-            if q1.real_field:
-                if q2.is_complex:
+            if q1.real_field and q2.is_complex:
                     return Quaternion(re(q2) + q1.a, im(q2) + q1.b, q1.c, q1.d)
-                else:
-                    # q2 is something strange, do not evaluate:
-                    return Add(q1, q2)
-            else:
+            elif q2.is_commutative:
                 return Quaternion(q1.a + q2, q1.b, q1.c, q1.d)
+            else:
+                # q2 is something strange, do not evaluate:
+                return Add(q1, q2)
```

Comment:
Maybe ValueError should be raised. (Otherwise the result would be even more strange.)

Reply:
Good idea; I’ll do it. Thanks for reviewing.

Reply:
I did the changes. For some reason, this behaviour, i.e. returning unevaluated Mul objects when adding a non-commutative expression, was tested explicitly in the module. But I agree with you that it is much better to raise a ValueError, so I added that and I removed those tests.